### ISSUE TITLE
Cannot launch mpirun on raspberry pi from Mac ( tcp_peer_recv_connect_ack: received unexpected process identifier)

### ISSUE THREAD

**Issue by loregen** [open-mpi/ompi#12240]
Please submit all the information below so that we can understand the working environment that is the context for your question.

* If you have a problem building or installing Open MPI, [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-building-or-installing-open-mpi).
* If you have a problem launching MPI or OpenSHMEM applications, [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-launching-mpi-or-openshmem-applications).
* If you have a problem running MPI or OpenSHMEM applications (i.e., after launching them), [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-running-mpi-or-openshmem-applications).

## Background information

For educational purposes I am trying to launch a MPI application from my MacBook and make it run on my Raspberry Pi.

### What version of Open MPI are you using? (e.g., v4.1.6, v5.0.1, git branch name and hash, etc.)

I am using Open MPI v4.1.5 on the "master" node (MacBook) and Open MPI v4.1.0 on the "slave" node (RaspberryPi).

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

On the raspberry I installed OpenMPI with "apt install openmpi-bin". It automatically installed version 4.1.0.
On my Mac I already had a copy of openmpi installed via homebrew (v5.0.1), but because I wanted a more compatible version with the one on the raspberry, I built another copy of openmpi 4.1.5 from source. I tried some other versions (4.1.0, 4.1.2, 4.1.4) , but I had problems during compilation. Version 4.1.5 is the nearest to 4.1.0 that I am able to compile on my MacBook. I tried compiling both with clang and homebrew gcc, and the issue is still there. 
During configure I passed --enable-heterogeneous flag. 

### Please describe the system on which you are running

* Operating system/version: 
1) DietPi v8.25.1
2) macOS Sonoma 14.0

* Computer hardware:
1) RPi 4 Model B (armv7l)
2) MacBook Air M1 2020
* Network type:
Home LAN 
-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

The two OpenMPI installations work by themselves. I can launch mpi applications both on Mac and Raspberry (on the raspberry I get "no protocol specified" but it works ).The problem arises when I try to run, from my Mac, the following command:
```shell
mpirun  --host 10.0.0.145 ./hello_mpi_armv7l
```

where 10.0.0.145 is the ip of the raspberry on my local network. 

I created two users called mpiuser on both machines, enabled ssh passwordless authentication (I can ssh from my Mac into the raspberry with mpiuser without needing password) and created a shared directory, which is mounted with nfs. 
hello_mpi_armv7l is an executable in the shared folder compiled for the raspberry. I also have a version compiled for Mac, as the ultimate goal was to run the same executable on both machines and make them communicate, but for now I can't even launch the armv7l executable from the Mac. 
Ivy executing the above command I get the following output:
```
No protocol specified
No protocol specified
No protocol specified
No protocol specified
[DietPi:24672] [[58140,0],1] tcp_peer_recv_connect_ack: received unexpected process identifier [[0,7395],0] from [[58140,0],0]
--------------------------------------------------------------------------
ORTE was unable to reliably start one or more daemons.
This usually is caused by:

* not finding the required libraries and/or binaries on
  one or more nodes. Please check your PATH and LD_LIBRARY_PATH
  settings, or configure OMPI with --enable-orterun-prefix-by-default

* lack of authority to execute on one or more specified nodes.
  Please verify your allocation and authorities.

* the inability to write startup files into /tmp (--tmpdir/orte_tmpdir_base).
  Please check with your sys admin to determine the correct location to use.

*  compilation of the orted with dynamic libraries when static are required
  (e.g., on Cray). Please check your configure cmd line and consider using
  one of the contrib/platform definitions for your system type.

* an inability to create a connection back to mpirun due to a
  lack of common network interfaces and/or no route found between
  them. Please check network connectivity (including firewalls
  and network routing requirements).
--------------------------------------------------------------------------
```

running with "--mca plm_base_verbose" I get the following additional output:
```
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: registering framework plm components
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: found loaded component rsh
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: component rsh register function successful
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: found loaded component slurm
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: component slurm register function successful
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: found loaded component isolated
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_register: component isolated has no register or open function
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: opening plm components
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: found loaded component rsh
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: component rsh open function successful
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: found loaded component slurm
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: component slurm open function successful
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: found loaded component isolated
[Lorenzos-MacBook-Air-2.local:38321] mca: base: components_open: component isolated open function successful
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select: Auto-selecting plm components
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Querying component [rsh]
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Query of component [rsh] set priority to 10
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Querying component [slurm]
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Querying component [isolated]
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Query of component [isolated] set priority to 0
[Lorenzos-MacBook-Air-2.local:38321] mca:base:select:(  plm) Selected component [rsh]
[Lorenzos-MacBook-Air-2.local:38321] mca: base: close: component slurm closed
[Lorenzos-MacBook-Air-2.local:38321] mca: base: close: unloading component slurm
[Lorenzos-MacBook-Air-2.local:38321] mca: base: close: component isolated closed
[Lorenzos-MacBook-Air-2.local:38321] mca: base: close: unloading component isolated
[Lorenzos-MacBook-Air-2.local:38321] [[58127,0],0] plm:rsh: final template argv:
	/opt/homebrew/bin/ssh <template>  orted -mca ess "env" -mca ess_base_jobid "3809411072" -mca ess_base_vpid "<template>" -mca ess_base_num_procs "2" -mca orte_node_regex "Lorenzos-MacBook-Air-[1:2],[2:10].0.0.145@0(2)" -mca orte_hnp_uri "3809411072.0;tcp://10.0.0.8:57528" --mca plm_base_verbose "100" -mca plm "rsh" --tree-spawn -mca routed "radix" -mca orte_parent_uri "3809411072.0;tcp://10.0.0.8:57528" -mca pmix "^s1,s2,cray,isolated"
No protocol specified
No protocol specified
[DietPi:25262] mca: base: components_register: registering framework plm components
[DietPi:25262] mca: base: components_register: found loaded component rsh
[DietPi:25262] mca: base: components_register: component rsh register function successful
[DietPi:25262] mca: base: components_open: opening plm components
[DietPi:25262] mca: base: components_open: found loaded component rsh
[DietPi:25262] mca: base: components_open: component rsh open function successful
[DietPi:25262] mca:base:select: Auto-selecting plm components
[DietPi:25262] mca:base:select:(  plm) Querying component [rsh]
[DietPi:25262] mca:base:select:(  plm) Query of component [rsh] set priority to 10
[DietPi:25262] mca:base:select:(  plm) Selected component [rsh]
No protocol specified
No protocol specified
[DietPi:25262] [[58127,0],1] tcp_peer_recv_connect_ack: received unexpected process identifier [[0,4067],0] from [[58127,0],0]
[DietPi:25262] mca: base: close: component rsh closed
[DietPi:25262] mca: base: close: unloading component rsh
--------------------------------------------------------------------------
```

**Comment by ggouaillardet** [comment]
Too many things can go wrong here:
- you are mixing two versions (`4.1.0` and `4.1.5`)
- the name suggests RPi is running 32 bits binaries, and I assume your Mac is 64 bits
- RPi is definitely Arm, but what about your Mac?

FWIW, the process identifier `[[0,4067],0]` suggests an endian-ness issue, but I am not quite sure about that.

I suspect the bare minimum is to double check Open MPI was configure'd with `--enable-heterogeneous` on both nodes.
That being said, I would start by installing a 64 bits OS on the RPi4, and then Open MPI 4.1.5.

**Comment by loregen** [comment]
@ggouaillardet 
Thank you very much for your quick response. 
I switched version on the Raspberry Pi. I built v4.1.5 (same as Mac) from source with --enable-heterogeneous.
Same error message appears. 
RPi is running 32 bit OS at the moment, while Mac is ARM 64 bit. 
My question is : is there a chance to make it work if the word lengths are different between the nodes? I know for a fact that different architectures is feasible : my friend managed to make it work between x86-64 and arm, but both the OS's were 64 bit in his case.

**Comment by ggouaillardet** [comment]
TL;DR I don't know.

When you say "same error message", can you confirm it reports an issue with process identifier `[[0,xyz],0]` which suggests an endianness issue?
On second thought, both M1 and RPi are little endian, so the root cause could be a different one.

I used to test it between Intel x86_64 (little endian 64 bits) and Sparc (big endian 64 bits) but that was quite a while ago.
I do not think I ever tested between 32 and 64 bits, and I am unaware of end users using heterogeneous systems in production.


The code for conversion is still there (`MCA_OOB_TCP_HDR_{NTOH,HTON}` is `orte/mca/oob/tcp/oob_tcp_connection.c`)

The puzzling part is is `jobid` and `vpid` have been swapped as if `{jobid,vpid}` was swapped as a 64 bit word, but these are swapped as two independent 32 bits unsigned int.

you can run with `mpirun --mca oob_base_verbose 100 ...` to see if we get some more useful logs, otherwise you can add some `printf()`, for example in `mca_oob_tcp_peer_recv_connect_ack()`, print `hdr.origin.jobid` and `hdr.origin.vpid` before and after `MCA_OOB_TCP_HDR_NTOH`, we might spot something obvious.

An other option would be to try an older Open MPI version (for example the v3 series), the root cause could be a bug that went undetected since basically nobody is testing this kind of environment.
